### ISSUE TITLE
Async/callback based interface between Span/Log Processor and Span/Log Exporter

### ISSUE THREAD

**Issue by lalitb** [open-telemetry/opentelemetry-cpp#1239]
As of now, the flow of spans from Span Processor to Exporter is synchronous, which means the processor has to wait for the upload/export status of the ongoing request before invoking a new request. 

``Application -> Span::OnEnd() -> SimpleProcessor::OnEnd -> Mutex Lock -> Exporter::Export()``

The application has to wait for the span to be uploaded to continue further with business logic. 

 The Batch processor simplifies this by batching/caching the spans in the processor and uploading them in a separate thread.
```
Application -> Span::OnEnd() -> BatchProcessor::OnEnd() -> CircularBuffer:Add(span)
            -> Span::OnEnd() -> BatchProcessor::OnEnd() -> CircularBuffer:Add(span)
            -> Span::OnEnd() -> BatchProcessor::OnEnd() -> CircularBuffer:Add(span)   
                                        ProcessorThread-> CircularBuffer::Drain() -> Export::Export()
Application -> Span::OnEnd() -> BatchProcessor::OnEnd() -> CircularBuffer:Add(span)  
```
This allows the Application to continue processing business logic without waiting for spans to export, and also Span::OnEnd() will not block even when export is ongoing in a separate thread ( there is no lock due to the use of  CircularBuffer to store Spans).

With BatchProcessor, while the interface between application and processor is fast (no wait), the interface between processor and exporter is still synchronous. As the processor needs to wait for the ongoing span(s) export to finish before invoking new export. This causes a bottleneck if the application is creating spans faster than the time taken by the exporter to export the spans, and eventually, the CircularBuffer gets full. Once the CircularBuffer gets full, Span::OnEnd() needs to wait/block on the buffer to be available for writing.

The current design for the interface between processor and exporter is driven by [specification ](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#span-exporter) as below
 - Export() will never be called concurrently for the same exporter instance. Export() can be called again only after the current call returns.
 - Success - The batch has been successfully exported. For protocol exporters this typically means that the data is sent over the wire and delivered to the destination server.
 - Note: this result may be returned via an async mechanism or a callback, if that is idiomatic for the language implementation.

So, the solution could be to return the exporter::Export() result as a callback or async mechanism,
 - Async mechanism: C++ provide async mechanism using std::async/std::promise. It would be good to do more research if its is suitable for more complex scenarios, and whether all the compiler provide a stable implementation for it before deciding on it ( Good read - https://www.cppstories.com/2014/01/tasks-with-stdfuture-and-stdasync/ )
 - Callback mechanism: would be much simpler something like:

```cpp
     bool Exporter:Export(const nostd::span<std::unique_ptr<Recordable>> &spans,
          nostd::function_ref<bool(ExportResult)> result_callback);
```
This allows the exporter to invoke the result_callback when the export status is available, and the processor to invoke the export without waiting for the ongoing export to complete.

Also, the interface change should be backward compatible, and shouldn't break the existing/external exporters. One option would be to have the processor decide which export mechanism to use based on the configurable parameter. The default should be as existing ( i.e, synchronous export).

**Comment by lalitb** [comment]
To start with, this new interface can be quickly integrated with Zipkin, and OTLP HTTP exporter, as the existing ext::http::client support [callback ](https://github.com/open-telemetry/opentelemetry-cpp/blob/11a219c80a7a3c47f479ce9d20e590e5900d0248/ext/include/opentelemetry/ext/http/client/http_client.h#L215)mechanism.

**Comment by lalitb** [comment]
@DebajitDas Also, you may want to look into a related PR #63, which was an attempt to add async support using [libevent](https://[libevent](https://libevent.org/).org/). The PR adds support for files i/o and timers and networking was to follow afterward. Somehow it was never merged for various reasons. 

Feel free to decide on the best approach on adding the async support :)

**Comment by DebajitDas** [comment]
Thanks @lalitb . I am looking at C++ async mechanism and see if that can be used to achieve asynchronous Export Call. 
I looked at http::client callback mechanism, but that seems to be blocking call, even though it was implemented with async mechanism. Am I missing something?

**Comment by owent** [comment]
I create another issue #1243 to implement async send for `opentelemetry::ext::http::client::curl::HttpClient`. But it's just for curl and http client.
To my understanding, there is no standard async mechanism in C++, `select()` can be used on all platforms we support but it has performence problem and it can not handle more than 1024 file descriptors(#1220 ) . Many libraries and system use [libevent][1], [libuv][2] or [boost.asio] as their event poller. (libcurl support use both [libevent][1] and [libuv][2] or use internal `select()` as fallback.). [hiredis][5] support more event management libraries or frameworks.

Some framework implement their own event and timer management(e.g. [redis][4]). It need to implement different poller for different system, IOCP for Windows, epoll for linux(maybe io_uring for modern linux kernel), kqueue for BSD, evport for Solaris and etc.

I'm thinking to use [`curl_multi_poll` ](https://curl.se/libcurl/c/curl_multi_poll.html) as the async mechanism in `opentelemetry::ext::http::client::curl::HttpClient`. Of course, It's better if we have a our own event system and all modules can use the same async mechanism.

[1]: https://libevent.org/
[2]: https://libuv.org/
[3]: https://www.boost.org/libs/asio/
[4]: https://redis.io/
[5]: https://github.com/redis/hiredis/tree/master/adapters

**Comment by lalitb** [comment]
> I looked at http::client callback mechanism, but that seems to be blocking call, even though it was implemented with async mechanism. Am I missing something?

@DebajitDas You are right. The abstract method [http::client::Session::SendRequest](https://github.com/open-telemetry/opentelemetry-cpp/blob/11a219c80a7a3c47f479ce9d20e590e5900d0248/ext/include/opentelemetry/ext/http/client/http_client.h#L215) takes a function object as an argument, and it's curl implementation [http::client::curl::Session::SendRequest](https://github.com/open-telemetry/opentelemetry-cpp/blob/11a219c80a7a3c47f479ce9d20e590e5900d0248/ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h#L145) is synchronous i.e, this method returns only after function object is called. This should probably change once @owent implements #1243. But we can only take full advantage of this change once the Exporter::export() also takes callback/function object. 

> To my understanding, there is no standard async mechanism in C++

@owent agree on this. Any async mechanism requires an event loop to run, and this is missing in C++. std::async potentially uses a separate thread to run the operation, and invoke the callback which is not a true async mechanism. That is the reason PR #63 to use libevent should be something we can focus more. It would be also useful to use this for metrics implementation, where we need to poll for metrics from multiple instruments simultaneously. What are your thoughts on using libevent?

**Comment by owent** [comment]
> What are your thoughts on using libevent?

In my opnion, we should let the user decide which event loop to use. Should we also implement a adaptor just like [hiredis](https://github.com/redis/hiredis/tree/master/adapters)? And we can use libevent by default.

**Comment by owent** [comment]
I also use [libwebsockets][1] in our project, it alse support to use [libuv][2] , [libevent][3], [libev][5], [libubox/uloop][4] and etc. as   event loop. I think [libuv][2]  and [libevent][3] is widely used now.

[1]: https://github.com/warmcat/libwebsockets/tree/main/lib/event-libs
[2]: https://libuv.org/
[3]: https://libevent.org/
[4]: https://git.openwrt.org/?p=project/libubox.git
[5]: https://github.com/enki/libev

**Comment by lalitb** [comment]
> In my opnion, we should let the user decide which event loop to use. Should we also implement a adaptor just like [hiredis](https://github.com/redis/hiredis/tree/master/adapters)? And we can use libevent by default.

This is also what #63 is trying to achieve. Provides an abstract adaptor, with libevent implementation.

**Comment by DebajitDas** [comment]
From the interface perspective of Processor ---> Exporter, callback mechanism could be achieved  by the following:
1. _Using std::async and std::shared_future_:  On every call to Export, Exporter would use std::async (std::launch::async) to launch a separate thread and perform the export work there. And when the result is ready, invoke the callback.
    std::shared_future can be stored in queue and a separate worker thread would run in exporter to fetch the shared_future to wait on the task to complete. 
    **Caveat:**  Every call to export creates a new thread, hence there might be performance penalty and system thread limit might be reached.
    Also, according to [this](https://ddanilov.me/std-async-implementations/) there is difference in implementation of GCC/LLVM and MSVC compilers. MSVC compiler uses threads from windows thread pool, therefore should carefully use thread_local variable as they might not be initialised.
    
 2.  The other option would be to launch thread_pool in the exporter, where worker threads would pick up its work and perform export and would invoke the callback on export completion.
And the number of threads could be kept configurable. In case libevent is used, I think one worker thread is sufficient. 

I might be wrong in my understanding and not at par with the context. Please correct me if I am missing something.

**Comment by lalitb** [comment]
I would prefer to offload most of the complexity of thread/concurrency management to Span/Log Processor and let the exporter logic be focused on the serialization and transport. This will also ease the work for exporter developers.

If we can scope this issue for only interface change, and think about concurrency separately, the rough changes for [ElasticsearchLogExporter ](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/exporters/elasticsearch/src/es_log_exporter.cc) would be (not tested):

```cpp
class ResponseHandler : public http_client::EventHandler
{
public:
  /**
   * Creates a response handler, that by default doesn't display to console
   */
  ResponseHandler(bool console_debug = false, 
        nostd::function_ref<bool(ExportResult)> result_callback, 
        std::shared_ptr<Session> &&session ) : 
	console_debug_{console_debug}, session_{std::move(session)}
  {
  }

  /**
   * Automatically called when the response is received, store the body into a string and notify any
   * threads blocked on this result
   */
  void OnResponse(http_client::Response &response) noexcept override
  {
	// Log error if response body has error
	session->FinishSession();
      result_callback(sdk::common::ExportResult::kSuccess);
	
  }

  void OnEvent(http_client::SessionState state, nostd::string_view reason) noexcept override
 {
     session->FinishSession();
     result_callback(sdk::common::ExportResult::kFailure)
 }
};


void ElasticsearchLogExporter::Export(
    const nostd::span<std::unique_ptr<sdklogs::Recordable>> &records,
nostd::function_ref<bool(ExportResult)> result_callback) noexcept
{

  // Create a connection to the ElasticSearch instance
  auto session = http_client_->CreateSession(options_.host_ + std::to_string(options_.port_));
  auto request = session->CreateRequest();
  request->SetUri(options_.index_ + "/_bulk?pretty");
  request->SetMethod(http_client::Method::Post);
  request->AddHeader("Content-Type", "application/json");
  request->SetTimeoutMs(std::chrono::milliseconds(1000 * options_.response_timeout_));

   body_vec = serialise(records); 
  request->SetBody(body_vec);
  std::unique_ptr<ResponseHandler> handler(new ResponseHandler(options_.console_debug_, result_callback, session));
  session->SendRequest(*handler);

}
```

Again, would be good to have more thoughts before finalizing, and also can discuss in the community meeting.

**Comment by owent** [comment]
> I would prefer to offload most of the complexity of thread/concurrency management to Span/Log Processor and let the exporter logic be focused on the serialization and transport.

Agree. Can we also provide a internal event loop and let the exporters decide whether to use it or spawn a new thread to finish 
 IO?

**Comment by lalitb** [comment]
> Agree. Can we also provide a internal event loop and let the exporters decide whether to use it or spawn a new thread to finish
IO?

Yes, we definitely need to have an internal event loop. In the case of HTTP exporter, the decision of using the event loop for IO should be with ext::http::client implementation. So if the ext::http::curl want to use curl_multi_poll it should use the event loop. Do you think it would be beneficial for non-HTTP exporters?

**Comment by owent** [comment]
> Yes, we definitely need to have an internal event loop. In the case of HTTP exporter, the decision of using the event loop for IO should be with ext::http::client implementation. So if the ext::http::curl want to use curl_multi_poll it should use the event loop if available. Do you think it would be beneficial for non-HTTP exporters?

Yes. And if other exporters use any SDK which allow users to get the raw socket fd. We can also use our internal event loop to handle it.

**Comment by DebajitDas** [comment]
> ```c++
> std::unique_ptr<ResponseHandler>
> ```

@lalitb Thanks for clarifying the scope. With the above, we need to maintain such that RequestHandler object remains valid even when callback is called in separate thread.

**Comment by lalitb** [comment]
>  With the above, we need to maintain such that RequestHandler object remains valid even when callback is called in a separate thread.

It's a good point. Probably we need to pass the ownership of ResultHandler from exporter to HttpClient, and let it clean up after calling. This would need changes in HTTP client library

> Should we maintain two different export versions - both with and without callback?

Yes, we need to do that for backward compatibility. Though maintaining ABI compatibility would be tricky, as any external exporter (built against the old SDK) would no longer work with the new SDK, and would need recompilation.

> Processor decides which version to call depending on configurable parameters.

True.

**Comment by DebajitDas** [comment]
Thanks @lalitb for clarifying the doubts.
Now from the implementation perspective, we need to provide implementation for those exporters which uses http_client_curl internally, such as elastic_log_exporter, otlp_http_exporter, otlp_http_log_exporter and zipkin exporter. All other log/span exporter would have empty implementation in the callback version of export call?

**Comment by owent** [comment]
After the new `Export` interface is done. I can modify #1209 to make otlp_http_exporter and otlp_http_log_exporter to use the new callback version `Export` call then.

**Comment by lalitb** [comment]
> Agree. Can we also provide a internal event loop and let the exporters decide whether to use it or spawn a new thread to finish

Have created #1250 for supporting the event loop framework. Feel free to modify the issue, or add more data there.

**Comment by DebajitDas** [comment]
As pointed out in [https://github.com/open-telemetry/opentelemetry-specification/issues/2434#issuecomment-1076997385]
I wanted to have the discussion around "max_async_export" which would be configurable parameter. 
1. What should be the default value of this variable? 
2. What should be the behaviour if the value is set as Zero (0) ? Since this holds no meaning, should this is re-set to default.
3. What should be the max value allowed for this parameter?

**Comment by owent** [comment]
> As pointed out in [https://github.com/[open-telemetry/opentelemetry-specification/issues/2434](https://github.com/open-telemetry/opentelemetry-specification/issues/2434)#issuecomment-1076997385] I wanted to have the discussion around "max_async_export" which would be configurable parameter.
> 
> 1. What should be the default value of this variable?
> 2. What should be the behaviour if the value is set as Zero (0) ? Since this holds no meaning, should this is re-set to default.
> 3. What should be the max value allowed for this parameter?

I use the variable name `max_concurrent_requests` which is mentioned in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc-concurrent-requests in `OtlpHttpExporter` and `OtlpHttpLogExporter` now. But I didn't find the specification about the default value. Current default value ( `8` ) in `OtlpHttpExporter` and `OtlpHttpLogExporter` is just based on my local benchmark and may cost about 70% CPU time at most without droping data.It may take less cost after we use multi_handle to implement `curl::HttpClient` in thfuture.

What's your suggestion about the default value?

**Comment by DebajitDas** [comment]
> What's your suggestion about the default value?

Yes, I had that in mind about the default value used in max_concurrent_requests and to start with we can go with 8 as default value.

**Comment by owent** [comment]
According to the latest https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md , Export() will never be called concurrently.
Could this issue be closed as #1413 is already merged?

**Comment by lalitb** [comment]
#1413 fixes this.
